adapter/src/main/kotlin/org/javacs/ktda/jdi/launch/KDACommandLineLauncher.kt

```diff
@@ -0,0 +1,208 @@
+
+        defaultArguments[ARG_CWD] = StringArgument(ARG_CWD, description = "Current working directory")
+
+        defaultArguments[ARG_ENVS] = StringArgument(ARG_ENVS, description = "Environment variables")
+
+        // Load TransportService 's implementation
+        transportService = Class.forName("com.sun.tools.jdi.SocketTransportService").getDeclaredConstructor().newInstance() as TransportService
+
+        if(transportService == null){
+            throw IllegalStateException("Failed to load com.sun.tools.jdi.SocketTransportService")
+        }
+
+    }
+
+    override fun name(): String {
+        return this.javaClass.name
+    }
+
+    override fun description(): String {
+        return "A custom launcher supporting cwd and env variables"
+    }
+
+    override fun defaultArguments(): Map<String, Connector.Argument> {
+        return this.defaultArguments
+    }
+
+    override fun toString(): String {
+        return name()
+    }
+
+    override fun transport(): Transport {
+        return transport
+    }
+
+    protected fun getOrDefault(arguments: Map<String, Connector.Argument>, argName: String): String {
+        return arguments[argName]?.value() ?: defaultArguments[argName]?.value() ?: ""
+    }
+
+    /**
+     * A customized method to launch the vm and connect to it, supporting cwd and env variables
+     */
+    @Throws(IOException::class, IllegalConnectorArgumentsException::class, VMStartException::class)
+    override fun launch(arguments: Map<String, Connector.Argument>): VirtualMachine {
+        val vm: VirtualMachine
+
+        val home = getOrDefault(arguments, ARG_HOME)
+        val options = getOrDefault(arguments, ARG_OPTIONS)
+        val main = getOrDefault(arguments, ARG_MAIN)
+        val suspend = getOrDefault(arguments, ARG_SUSPEND).toBoolean()
+        val quote = getOrDefault(arguments, ARG_QUOTE)
+        var exe = getOrDefault(arguments, ARG_VM_EXEC)
+        val cwd = getOrDefault(arguments, ARG_CWD)
+        val envs = urlDecode(getOrDefault(arguments, ARG_ENVS))?.toTypedArray()
+
+        check(quote.length == 1) {"Invalid length for $ARG_QUOTE: $quote"}
+        check(!options.contains("-Djava.compiler=") ||
+                options.toLowerCase().contains("-djava.compiler=none")) { "Cannot debug with a JIT compiler. $ARG_OPTIONS: $options"}
+
+        val listenKey = transportService?.startListening() ?: throw IllegalStateException("Failed to do transportService.startListening()")
+        val address = listenKey.address()
+
+        try {
+            val command = StringBuilder()
+
+            exe = if (home.isNotEmpty()) Paths.get(home, "bin", exe).toString() else exe
+            command.append(wrapWhitespace(exe))
+
+            command.append(" $options")
+
+            //debug options
+            command.append(" -agentlib:jdwp=transport=${transport.name()},address=$address,server=n,suspend=${if (suspend) 'y' else 'n'}")
+
+            command.append(" $main")
```

Comment:
Same here, I think it looks a bit nicer with less whitespace.

```suggestion
            command.append(wrapWhitespace(exe))
            command.append(" $options")
            // debug options
            command.append(" -agentlib:jdwp=transport=${transport.name()},address=$address,server=n,suspend=${if (suspend) 'y' else 'n'}")
            command.append(" $main")
```